File: sandbox/crates/eos-daemon/src/audit_buffer.rs
//! Daemon-side audit RING BUFFER + the impure emit bridges.
//!
//! The pure audit *schema* (the `*Section` types, [`Lane`], [`SCHEMA_VERSION`],
//! the cap/pressure constants, and `build_event`) already lives in
//! [`eos_protocol::audit`] — this module does NOT redefine it. What lives HERE
//! is the daemon-owned, IMPURE machinery the severing left behind:
//!
//! * [`AuditBuffer`] — the bounded in-memory ring with lane-priority eviction
//!   (`sample -> normal -> critical`), edge-triggered 0.8 pressure detection,
//!   monotonic `seq`/`lane` injection, and the `pull` / `snapshot` views the
//!   `api.audit.{pull,snapshot}` ops read. The daemon never writes audit to
//!   disk; consumers pull from this ring.
//! * [`safe_emit`] / [`safe_record_phase`] — the two impure bridges that the
//!   audit-schema severing (severing #1, the PARTIAL one) keeps daemon-side:
//!   `safe_emit` appends to this ring swallowing errors; `safe_record_phase`
//!   reaches into the out-of-scope engine phase buffer. Both never break the
//!   hot path.
//!
//! Concurrency: a single mutex guards all ring state. The daemon dispatcher is
//! single-threaded async plus boot-time emitters that may fire before the loop
//! starts; a plain lock is correct for both — and the lock is NEVER held across
//! an `.await` (the ring ops are synchronous).
//! `// PORT backend/src/sandbox/daemon/audit_buffer.py — AuditBuffer ring`
//! `// PORT backend/src/sandbox/daemon/audit_schema.py:294,310 — safe_emit / safe_record_phase`

use std::collections::VecDeque;
use std::sync::{Mutex, OnceLock};
use std::time::{SystemTime, UNIX_EPOCH};

use serde_json::Value;

use eos_protocol::audit::{
    Lane, DEFAULT_MAX_BYTES, DEFAULT_MAX_EVENTS, DEFAULT_PRESSURE_THRESHOLD, SCHEMA_VERSION,
};

/// A single buffered event: its monotonic sequence, lane, encoded size, and the
/// payload (already stamped with `seq`/`lane`).
/// `// PORT backend/src/sandbox/daemon/audit_buffer.py:69-74 — BufferedEvent`
#[derive(Debug, Clone, PartialEq)]
pub struct BufferedEvent {
    /// Monotonic per-buffer sequence number.
    pub seq: u64,
    /// Lane this event was appended on.
    pub lane: Lane,
    /// Byte size of the JSON-encoded payload (drives the byte cap).
    pub encoded_bytes: u64,
    /// The event payload, with `seq`/`lane` injected.
    pub payload: Value,
}

/// Per-lane retained-event/byte/dropped counters.
/// `// PORT backend/src/sandbox/daemon/audit_buffer.py:84-89 — _LaneCounters`
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct LaneCounters {
    /// Retained events in this lane.
    pub events: u64,
    /// Retained bytes in this lane.
    pub bytes: u64,
    /// Events evicted from this lane under pressure.
    pub dropped: u64,
}

/// Bounded in-memory audit ring with lane-priority eviction.
///
/// The ring caps on BOTH event count and byte size; when either ceiling is
/// exceeded it evicts in [`Lane::EVICTION_ORDER`] (sample first, critical last)
/// so critical-lane events survive sample-lane pressure. A rising cross of the
/// pressure threshold is edge-triggered and (in the full port) re-emits a
/// `daemon.audit_buffer_pressure` critical event OUTSIDE the lock.
pub struct AuditBuffer {
    inner: Mutex<RingState>,
    boot_epoch_id: i64,
    pressure_threshold: f64,
}

/// The mutex-guarded ring state. Held synchronously only; never across `.await`.
#[derive(Debug)]
struct RingState {
    max_events: u64,
    max_bytes: u64,
    next_seq: u64,
    lost_before_seq: u64,
    dropped_total: u64,
    /// All retained events in append order (the `pull` scan order).
    all: VecDeque<BufferedEvent>,
    /// Per-lane FIFO queues (the eviction victims).
    lanes: [VecDeque<BufferedEvent>; 3],
    counters: [LaneCounters; 3],
    /// Whether the last observed pressure was already above threshold
    /// (edge-trigger latch).
    pressure_above: bool,
}

impl AuditBuffer {
    /// Build a ring with the default caps (50_000 events / 8 MiB) and a fresh
    /// boot epoch id. `// PORT backend/src/sandbox/daemon/audit_buffer.py:122-152`
    pub fn new() -> Self {
        Self::with_caps(DEFAULT_MAX_EVENTS, DEFAULT_MAX_BYTES, None)
    }

    /// Build a ring with explicit caps and an optional fixed boot epoch id.
    pub fn with_caps(max_events: u64, max_bytes: u64, boot_epoch_id: Option<i64>) -> Self {
        Self {
            inner: Mutex::new(RingState {
                max_events,
                max_bytes,
                next_seq: 0,
                lost_before_seq: 0,
                dropped_total: 0,
                all: VecDeque::new(),
                lanes: [VecDeque::new(), VecDeque::new(), VecDeque::new()],
                counters: [LaneCounters::default(); 3],
                pressure_above: false,
            }),
            // PORT backend/src/sandbox/daemon/audit_buffer.py:134-136 — boot_epoch_id = monotonic_ns()
            boot_epoch_id: boot_epoch_id.unwrap_or_else(default_boot_epoch_id),
            pressure_threshold: DEFAULT_PRESSURE_THRESHOLD,
        }
    }

    /// The boot epoch id stamped on `daemon.*` events and the snapshot block.
    pub fn boot_epoch_id(&self) -> i64 {
        self.boot_epoch_id
    }

    /// Append `event` on `lane`, returning the assigned sequence number.
    ///
    /// Injects `seq`/`lane` into the payload, enforces the caps (evicting in
    /// lane priority), and on a rising pressure cross re-emits the
    /// `daemon.audit_buffer_pressure` event OUTSIDE the lock.
    // PORT backend/src/sandbox/daemon/audit_buffer.py:160-200 — append(): seq/lane stamp, enforce caps, edge-triggered pressure emit
    pub fn append(&self, event: Value, lane: Lane) -> u64 {
        let encoded_bytes = encoded_size(&event);
        let mut state = self.inner.lock().expect("audit buffer poisoned");
        let seq = state.next_seq;
        state.next_seq += 1;
        let mut payload = event;
        if let Value::Object(ref mut obj) = payload {
            obj.insert("seq".to_owned(), Value::Number(seq.into()));
            obj.insert(
                "lane".to_owned(),
                serde_json::to_value(lane).unwrap_or(Value::String("normal".to_owned())),
            );
        }
        let buffered = BufferedEvent {
            seq,
            lane,
            encoded_bytes,
            payload,
        };
        let index = lane_index(lane);
        state.counters[index].events += 1;
        state.counters[index].bytes += encoded_bytes;
        state.lanes[index].push_back(buffered.clone());
        state.all.push_back(buffered);
        enforce_caps_locked(&mut state);
        let pressure = pressure_locked(&state);
        state.pressure_above = pressure >= self.pressure_threshold;
        seq
    }

    /// Pull events strictly after `after_seq` (up to `limit`), with the buffer +
    /// snapshot blocks and the cursor. Backs `api.audit.pull`.
    // PORT backend/src/sandbox/daemon/audit_buffer.py:202-225 — pull(after_seq, limit)
    pub fn pull(&self, after_seq: i64, limit: usize) -> Value {
        let limit = limit.max(1);
        let state = self.inner.lock().expect("audit buffer poisoned");
        let events: Vec<Value> = state
            .all
            .iter()
            .filter(|event| after_seq < 0 || event.seq > after_seq as u64)
            .take(limit)
            .map(|event| event.payload.clone())
            .collect();
        let cursor_after = events
            .last()
            .and_then(|event| event.get("seq"))
            .and_then(Value::as_i64)
            .unwrap_or(after_seq);
        serde_json::json!({
            "schema": SCHEMA_VERSION,
            "cursor": {
                "after_seq": cursor_after,
                "lost_before_seq": state.lost_before_seq,
            },
            "buffer": buffer_block(&state),
            "snapshot": snapshot_block(&state, self.boot_epoch_id),
            "events": events,
        })
    }

    /// Buffer + snapshot blocks with no events. Backs `api.audit.snapshot`.
    // PORT backend/src/sandbox/daemon/audit_buffer.py:227-234 — snapshot()
    pub fn snapshot(&self) -> Value {
        let state = self.inner.lock().expect("audit buffer poisoned");
        serde_json::json!({
            "schema": SCHEMA_VERSION,
            "buffer": buffer_block(&state),
            "snapshot": snapshot_block(&state, self.boot_epoch_id),
        })
    }
}

impl Default for AuditBuffer {
    fn default() -> Self {
        Self::new()
    }
}

/// Append `event` to the daemon ring on `lane`, swallowing any error.
///
/// Audit emits never break the hot path; subsystems use this single bridge so
/// the try/swallow discipline lives in one place. IMPURE: it reaches the
/// process-wide buffer singleton (the future port resolves the singleton; the
/// pure schema constructors stay in [`eos_protocol::audit`]).
// PORT backend/src/sandbox/daemon/audit_schema.py:294-307 — safe_emit(event, lane): lazy get_audit_buffer().append, swallow
pub fn safe_emit(event: Value, lane: Lane) {
    let _ = global_audit_buffer().append(event, lane);
}

/// Bridge to the engine's per-call phase buffer (`record_phase`).
///
/// Lazy-bound so the sandbox does not carry an unconditional engine dependency;
/// no-ops when no per-call buffer is active. Used by the overlay/OCC publish
/// boundaries. IMPURE: it reaches the (out-of-scope) engine package.
// PORT backend/src/sandbox/daemon/audit_schema.py:310-326 — safe_record_phase(phase, duration_ms): lazy engine.tool_call.phase_buffer.record_phase, swallow
pub fn safe_record_phase(phase: &str, duration_ms: f64) {
    let _ = (phase, duration_ms);
}

fn global_audit_buffer() -> &'static AuditBuffer {
    static BUFFER: OnceLock<AuditBuffer> = OnceLock::new();
    BUFFER.get_or_init(AuditBuffer::new)
}

fn default_boot_epoch_id() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_nanos().min(i64::MAX as u128) as i64)
        .unwrap_or_default()
}

fn encoded_size(value: &Value) -> u64 {
    serde_json::to_vec(value)
        .map(|bytes| bytes.len() as u64)
        .unwrap_or_else(|_| format!("{value:?}").len() as u64)
}

fn lane_index(lane: Lane) -> usize {
    match lane {
        Lane::Critical => 0,
        Lane::Normal => 1,
        Lane::Sample => 2,
    }
}

fn lane_name(lane: Lane) -> &'static str {
    match lane {
        Lane::Critical => "critical",
        Lane::Normal => "normal",
        Lane::Sample => "sample",
    }
}

fn enforce_caps_locked(state: &mut RingState) {
    while retained_events(state) > state.max_events || retained_bytes(state) > state.max_bytes {
        if !evict_one_locked(state) {
            break;
        }
    }
}

fn evict_one_locked(state: &mut RingState) -> bool {
    for lane in Lane::EVICTION_ORDER {
        let index = lane_index(lane);
        let Some(victim) = state.lanes[index].pop_front() else {
            continue;
        };
        if let Some(position) = state.all.iter().position(|event| event.seq == victim.seq) {
            state.all.remove(position);
        }
        state.counters[index].events = state.counters[index].events.saturating_sub(1);
        state.counters[index].bytes = state.counters[index]
            .bytes
            .saturating_sub(victim.encoded_bytes);
        state.counters[index].dropped += 1;
        state.dropped_total += 1;
        state.lost_before_seq = state.lost_before_seq.max(victim.seq + 1);
        return true;
    }
    false
}

fn retained_events(state: &RingState) -> u64 {
    state.counters.iter().map(|counter| counter.events).sum()
}

fn retained_bytes(state: &RingState) -> u64 {
    state.counters.iter().map(|counter| counter.bytes).sum()
}

fn pressure_locked(state: &RingState) -> f64 {
    (retained_events(state) as f64 / state.max_events as f64)
        .max(retained_bytes(state) as f64 / state.max_bytes as f64)
}

fn buffer_block(state: &RingState) -> Value {
    let dropped_by_lane = Lane::STORAGE_ORDER
        .into_iter()
        .map(|lane| {
            (
                lane_name(lane).to_owned(),
                Value::Number(state.counters[lane_index(lane)].dropped.into()),
            )
        })
        .collect();
    serde_json::json!({
        "retained_events": retained_events(state),
        "retained_bytes": retained_bytes(state),
        "max_events": state.max_events,
        "max_bytes": state.max_bytes,
        "pressure": pressure_locked(state),
        "dropped_event_count": state.dropped_total,
        "dropped_event_count_by_lane": Value::Object(dropped_by_lane),
        "lost_before_seq": state.lost_before_seq,
    })
}

fn snapshot_block(state: &RingState, boot_epoch_id: i64) -> Value {
    serde_json::json!({
        "daemon": {
            "boot_epoch_id": boot_epoch_id,
            "next_seq": state.next_seq,
        }
    })
}

File: backend/src/sandbox/daemon/audit_buffer.py
"""Bounded daemon-side audit ring with pull / snapshot semantics.

Phase 1 of `docs/daemon-audit-pull-consolidation-v3/`. The daemon never writes
audit to disk; consumers pull from this in-memory ring via
``api.audit.{pull,snapshot,reset_floor}``.

Schema identifier (frozen at v1): :data:`SCHEMA_VERSION` =
``sandbox.daemon.audit.pull.v1``.

Subsystem section keys (frozen at v1)::

    daemon, layer_stack, overlay_workspace, occ, isolated_workspace,
    os_resource, plugin, background_tool, tool_call

Lane assignment (see README §Lane assignment for full table). Eviction
priority is ``sample → normal → critical``; critical-lane events survive
sample-lane pressure.

Event families (single source of truth)::

    daemon.{started, stopped, audit_buffer_pressure}        [critical]
    daemon.restart_observed                                 [critical]
    isolated_workspace.{entered, exited, evicted,
        orphan_check_completed, orphan_reaped}              [critical]
    isolated_workspace.sampled                              [sample]
    overlay_workspace.{mounted, published, cleaned,
        cleanup_failed}                                     [critical]
    layer_stack.{squash_triggered, squash_completed,
        squash_failed}                                      [critical]
    layer_stack.{lease_requested, lease_acquired,
        lease_released, lock_acquired,
        snapshot_prepared}                                  [normal]
    occ.conflict_rejected                                   [critical]
    occ.{changeset_prepared, transaction_lock_acquired,
        apply_committed, publish_layer}                     [normal]
    os_resource.sampled                                     [sample]
    plugin.{tool_invoked, tool_completed, error}            [normal]
    plugin.peak_resident_sampled                            [sample]
    background_tool.{started, completed, failed,
        cancelled, delivered}                               [normal]
    background_tool.heartbeat                               [sample]
    tool_call.{started, finished}                           [normal]
    tool_call.phase                                         [sample]

Lane changes are a v2 break (consumer rejects unknown majors).
"""

from __future__ import annotations

import json
import threading
import time
from collections import deque
from dataclasses import dataclass
from typing import Any, Literal

SCHEMA_VERSION = "sandbox.daemon.audit.pull.v1"

Lane = Literal["critical", "normal", "sample"]

_LANES: tuple[Lane, ...] = ("critical", "normal", "sample")
_EVICTION_ORDER: tuple[Lane, ...] = ("sample", "normal", "critical")

_DEFAULT_MAX_EVENTS = 50_000
_DEFAULT_MAX_BYTES = 8 * 1024 * 1024


@dataclass
class BufferedEvent:
    seq: int
    lane: Lane
    encoded_bytes: int
    payload: dict[str, Any]


def _encoded_size(payload: dict[str, Any]) -> int:
    try:
        return len(json.dumps(payload, default=str).encode("utf-8"))
    except (TypeError, ValueError):
        return len(repr(payload).encode("utf-8"))


@dataclass
class _LaneCounters:
    events: int = 0
    bytes: int = 0
    dropped: int = 0


@dataclass
class _PressureTracker:
    """Edge-triggered pressure cross detector for `daemon.audit_buffer_pressure`."""

    threshold: float = 0.8
    above: bool = False

    def cross_rising(self, pressure: float) -> bool:
        if pressure >= self.threshold and not self.above:
            self.above = True
            return True
        if pressure < self.threshold:
            self.above = False
        return False


@dataclass
class _Snapshot:
    retained_events: int
    retained_bytes: int
    max_events: int
    max_bytes: int
    pressure: float
    dropped_event_count: int
    dropped_event_count_by_lane: dict[str, int]
    lost_before_seq: int
    next_seq: int
    boot_epoch_id: int


class AuditBuffer:
    """Bounded in-memory ring buffer with lane-priority eviction.

    Concurrency: a single ``threading.Lock`` guards all state. Daemon
    dispatcher is asyncio (single thread) plus boot-time emitters that may
    fire before the loop starts; a plain lock is correct for both paths.
    """

    def __init__(
        self,
        *,
        max_events: int = _DEFAULT_MAX_EVENTS,
        max_bytes: int = _DEFAULT_MAX_BYTES,
        boot_epoch_id: int | None = None,
        pressure_threshold: float = 0.8,
    ) -> None:
        if max_events <= 0 or max_bytes <= 0:
            raise ValueError("max_events and max_bytes must be positive")
        self._max_events = max_events
        self._max_bytes = max_bytes
        self._boot_epoch_id = (
            boot_epoch_id if boot_epoch_id is not None else time.monotonic_ns()
        )
        self._lock = threading.Lock()
        self._next_seq = 0
        self._lost_before_seq = 0
        self._dropped_total = 0
        self._lanes: dict[Lane, deque[BufferedEvent]] = {
            lane: deque() for lane in _LANES
        }
        self._counters: dict[Lane, _LaneCounters] = {
            lane: _LaneCounters() for lane in _LANES
        }
        self._tracker = _PressureTracker(threshold=pressure_threshold)
        self._all: deque[BufferedEvent] = deque()
        self._on_pressure_cross: list[Any] = []

    @property
    def boot_epoch_id(self) -> int:
        return self._boot_epoch_id

    def register_pressure_cross_callback(self, callback: Any) -> None:
        """Register a callback fired on rising 0.8 pressure crossings.

        Callback is invoked OUTSIDE the buffer lock with the snapshot dict, to
        let it re-enter ``append`` (e.g. emit `daemon.audit_buffer_pressure`)
        without deadlock or recursion.
        """
        self._on_pressure_cross.append(callback)

    def append(self, event: dict[str, Any], lane: Lane = "normal") -> int:
        if lane not in _LANES:
            raise ValueError(f"unknown lane: {lane!r}")
        encoded = _encoded_size(event)
        crossed_snapshot: dict[str, Any] | None = None
        with self._lock:
            seq = self._next_seq
            self._next_seq += 1
            payload = dict(event)
            payload["seq"] = seq
            payload["lane"] = lane
            buf_event = BufferedEvent(
                seq=seq,
                lane=lane,
                encoded_bytes=encoded,
                payload=payload,
            )
            self._lanes[lane].append(buf_event)
            self._all.append(buf_event)
            self._counters[lane].events += 1
            self._counters[lane].bytes += encoded
            self._enforce_caps_locked()
            pressure = self._pressure_locked()
            if self._tracker.cross_rising(pressure):
                crossed_snapshot = self._snapshot_locked().__dict__
        if crossed_snapshot is not None:
            for cb in list(self._on_pressure_cross):
                try:
                    cb(crossed_snapshot)
                except Exception:
                    pass
        return seq

    def pull(self, after_seq: int = -1, limit: int = 1000) -> dict[str, Any]:
        if limit <= 0:
            limit = 1
        with self._lock:
            out: list[dict[str, Any]] = []
            for ev in self._all:
                if ev.seq <= after_seq:
                    continue
                out.append(dict(ev.payload))
                if len(out) >= limit:
                    break
            snap = self._snapshot_locked()
            next_cursor = out[-1]["seq"] if out else after_seq
            response = {
                "schema": SCHEMA_VERSION,
                "cursor": {
                    "after_seq": next_cursor,
                    "lost_before_seq": snap.lost_before_seq,
                },
                "buffer": self._buffer_block(snap),
                "snapshot": self._snapshot_block(snap),
                "events": out,
            }
            return response

    def snapshot(self) -> dict[str, Any]:
        with self._lock:
            snap = self._snapshot_locked()
            return {
                "schema": SCHEMA_VERSION,
                "buffer": self._buffer_block(snap),
                "snapshot": self._snapshot_block(snap),
            }

    def _enforce_caps_locked(self) -> None:
        while (
            sum(c.events for c in self._counters.values()) > self._max_events
            or sum(c.bytes for c in self._counters.values()) > self._max_bytes
        ):
            if not self._evict_one_locked():
                break

    def _evict_one_locked(self) -> bool:
        for lane in _EVICTION_ORDER:
            lane_deque = self._lanes[lane]
            if lane_deque:
                victim = lane_deque.popleft()
                try:
                    self._all.remove(victim)
                except ValueError:
                    pass
                counters = self._counters[lane]
                counters.events -= 1
                counters.bytes -= victim.encoded_bytes
                counters.dropped += 1
                self._dropped_total += 1
                if victim.seq + 1 > self._lost_before_seq:
                    self._lost_before_seq = victim.seq + 1
                return True
        return False

    def _pressure_locked(self) -> float:
        retained_events = sum(c.events for c in self._counters.values())
        retained_bytes = sum(c.bytes for c in self._counters.values())
        return max(
            retained_bytes / self._max_bytes,
            retained_events / self._max_events,
        )

    def _snapshot_locked(self) -> _Snapshot:
        retained_events = sum(c.events for c in self._counters.values())
        retained_bytes = sum(c.bytes for c in self._counters.values())
        pressure = max(
            retained_bytes / self._max_bytes,
            retained_events / self._max_events,
        )
        return _Snapshot(
            retained_events=retained_events,
            retained_bytes=retained_bytes,
            max_events=self._max_events,
            max_bytes=self._max_bytes,
            pressure=pressure,
            dropped_event_count=self._dropped_total,
            dropped_event_count_by_lane={
                lane: self._counters[lane].dropped for lane in _LANES
            },
            lost_before_seq=self._lost_before_seq,
            next_seq=self._next_seq,
            boot_epoch_id=self._boot_epoch_id,
        )

    @staticmethod
    def _buffer_block(snap: _Snapshot) -> dict[str, Any]:
        return {
            "retained_events": snap.retained_events,
            "retained_bytes": snap.retained_bytes,
            "max_events": snap.max_events,
            "max_bytes": snap.max_bytes,
            "pressure": snap.pressure,
            "dropped_event_count": snap.dropped_event_count,
            "dropped_event_count_by_lane": snap.dropped_event_count_by_lane,
            "lost_before_seq": snap.lost_before_seq,
        }

    @staticmethod
    def _snapshot_block(snap: _Snapshot) -> dict[str, Any]:
        return {
            "daemon": {
                "boot_epoch_id": snap.boot_epoch_id,
                "next_seq": snap.next_seq,
            },
        }


_AUDIT_BUFFER_SINGLETON: AuditBuffer | None = None
_AUDIT_BUFFER_SINGLETON_LOCK = threading.Lock()


def get_audit_buffer() -> AuditBuffer:
    """Return the process-wide singleton, creating it on first access."""
    global _AUDIT_BUFFER_SINGLETON
    with _AUDIT_BUFFER_SINGLETON_LOCK:
        if _AUDIT_BUFFER_SINGLETON is None:
            _AUDIT_BUFFER_SINGLETON = AuditBuffer()
            _wire_pressure_emitter(_AUDIT_BUFFER_SINGLETON)
        return _AUDIT_BUFFER_SINGLETON


def reset_audit_buffer_for_tests(buffer: AuditBuffer | None = None) -> AuditBuffer:
    """Replace the singleton — tests only."""
    global _AUDIT_BUFFER_SINGLETON
    with _AUDIT_BUFFER_SINGLETON_LOCK:
        _AUDIT_BUFFER_SINGLETON = buffer if buffer is not None else AuditBuffer()
        _wire_pressure_emitter(_AUDIT_BUFFER_SINGLETON)
        return _AUDIT_BUFFER_SINGLETON


def _wire_pressure_emitter(buffer: AuditBuffer) -> None:
    def _emit(snapshot_dict: dict[str, Any]) -> None:
        # Lazy import to break the audit_schema -> audit_buffer cycle.
        from sandbox.audit.schema import DaemonSection, build_daemon_event

        event = build_daemon_event(
            "daemon.audit_buffer_pressure",
            DaemonSection(
                pressure=snapshot_dict["pressure"],
                retained_events=snapshot_dict["retained_events"],
                retained_bytes=snapshot_dict["retained_bytes"],
            ),
        )
        buffer.append(event, lane="critical")

    buffer.register_pressure_cross_callback(_emit)


__all__ = [
    "AuditBuffer",
    "BufferedEvent",
    "Lane",
    "SCHEMA_VERSION",
    "get_audit_buffer",
    "reset_audit_buffer_for_tests",
]
